lib/RestructureCFG/InlineDispatcherSwitch.cpp

```diff
@@ -737,13 +737,140 @@ inlineDispatcherSwitchImpl(ASTTree &AST,
+      If->setThen(NewThen);
+    }
+    if (If->hasElse()) {
+      ASTNode *Else = If->getElse();
+      ASTNode *NewElse = rc_recur
+        simplifySuperfluousNonLocalCFImpl(Else, FallThroughScopeMap);
+      If->setElse(NewElse);
+    }
+  } break;
+  case ASTNode::NK_Switch: {
+    auto *Switch = llvm::cast<SwitchNode>(Node);
+
+    // First of all, we recursively process the `case` nodes contained in the
+    // `switch` in order to process the inner portion of the AST
+    llvm::SmallVector<size_t> ToRemoveCaseIndex;
+    for (auto &Group : llvm::enumerate(Switch->cases())) {
+      unsigned Index = Group.index();
+      auto &LabelCasePair = Group.value();
+      LabelCasePair.second = rc_recur
+        simplifySuperfluousNonLocalCFImpl(LabelCasePair.second,
+                                          FallThroughScopeMap);
+
+      if (LabelCasePair.second == nullptr) {
+        ToRemoveCaseIndex.push_back(Index);
+      }
+    }
+
+    for (auto ToRemoveCase : llvm::reverse(ToRemoveCaseIndex)) {
+      Switch->removeCaseN(ToRemoveCase);
+    }
+  } break;
+  case ASTNode::NK_Code:
+  case ASTNode::NK_Set:
+  case ASTNode::NK_SwitchBreak:
+  case ASTNode::NK_Continue:
+  case ASTNode::NK_Break:
+    // Do nothing
+    break;
+  default:
+    revng_unreachable();
+  }
+
+  rc_return Node;
+}
+
+/// This simplification routine is aimed at removing superfluous
+/// `NonLocalControl Flow` statements that may remain around after the inline of
+/// `cases`. Specifically, if a `return` statement is moved into an inner loop
+/// in place of a `SetNode`, it may be that a following `break` statement can be
+/// simplified away.
+static ASTNode *simplifySuperfluousNonLocalCF(ASTTree &AST,
+                                              ASTNode *RootNode,
+                                              const FallThroughScopeTypeMap
+                                                &FallThroughScopeMap) {
+
+  // Call the underlying implementation
+  RootNode = simplifySuperfluousNonLocalCFImpl(RootNode, FallThroughScopeMap);
+  AST.setRoot(RootNode);
+
+  return RootNode;
+}
+
+ASTNode *inlineDispatcherSwitch(const model::Binary &Model,
+                                ASTTree &AST,
+                                ASTNode *RootNode) {
 
   // Compute the loop -> dispatcher switch correspondence
   LoopDispatcherMap LoopDispatcherMap = computeLoopDispatcher(AST);
 
   RootNode = inlineDispatcherSwitchImpl(AST, RootNode, LoopDispatcherMap);
 
+  // Pre-compute the `FallThroughScopeType` before the `SuperfluousNonLocalCF`
+  // simplidy is performed. This information is useful in order to decide
+  // whether we remove some statements that after the `case` inlining are
+  // preceded by `return` statements.
+  FallThroughScopeTypeMap
+    FallThroughScopeMap = computeFallThroughScope(Model, RootNode);
+
+  // Perform the `SuperfluousNonLocalCF` simplification pass
+  RootNode = simplifySuperfluousNonLocalCF(AST, RootNode, FallThroughScopeMap);
+
```

Comment:
This looks like a dead-code-elimination pass.
Please carve this out in a separate beautifier.
And give it a name like `removeDeadCode`.

Reply:
I implemented a `removeDeadCode` beautify pass in a separate commit 266c681abb4668354bbbbc5cb6802ad4c583ac30, so everything is more autocontained